### DISCUSSION TITLE
Original game engine bugs

### DISCUSSION THREAD

**Discussion by induktio** [induktio/thinker#29] [Ideas]
Scient listed some notable game engine bugs in [OpenSMACX issues](https://github.com/b-casey/OpenSMACX/issues). Latest Thinker version should fix most of the serious bugs/crashes in the original game.

The remaining issues here are more situational and/or there's no implemented code ready to fix them. For example it would be nice to have better pathfinding for all units, but the binary interface would need to be reversed fully first.

Fixes implemented in latest version:
* Crash or weird behavior when using 'Edit Map' to reset fungus
* Minor bug: Gender switching when clicking in base window
* Edge case where faction could get eliminated even when they had active colony pods
* Base names visible in unexplored tiles when triggering certain movement actions
* When gifting energy credits, response message (GAVEENERGY) isn't displayed
* Saved game can use incorrect faction colors
* SMAC mode in SMACX - Orbital attacks are broken
* Evicted sea triad probe teams can be returned to a landlocked base
* Pact AI poor decisions making when swapping bases
* Quick Start game has potential for game to freeze if Hive is one of factions
* Stack corruption in wants_to_attack()

Issues probably not relevant with Thinker:
* Coastal calculation with base_coast()
* AI base function incorrectly using negative queue id in logic
* AI Probe Teams keep attempting to Infiltrate Datalinks after success
* enemy_capabilities() issues

Issues not fixed yet:
* Bug: Goto-Suggestion ignores ZoC
* Saved zoom level lost when loading a game
* Scenario Editor undo/redo sometimes require multiple attempts
* Scenario Editor redo mechanic last change
* Inconsistent behavior with player_data mind control variables
* Certain Probe actions against base with extra security can abort action without penalty

Base window freeze issue (noted by John-Green223)
* Steps to reproduce: Open diplomacy with another faction > Click Profile > Click Base operations > Open any city. Game interface is now stuck on the base window and it cannot be closed.

Inconsistent behaviour with drone/psych calculation (noted by John-Green223)
* SE +3 police rating "police effect doubled" in combination with non-lethal methods ability suppresses only 3 drones per garrison unit instead of 4 like the description would imply.

Possible misleading diplomacy dialogue (noted by John-Green223)
* Sometimes allied AIs will attempt to trade commlinks for unknown factions for energy or techs (METFRIEND dialogue path), but when asked separately to give the commlinks just for goodwill and friendship, they will always agree regardless causing the trade decision to be useless.

Incorrect transport unit cargo handling
* Whenever multiple transports with some of them being loaded enter into the same tile, the carried units might get randomly shifted into another transport. While this happens for all factions, it is most problematic for AI move planning.

Balance or inconsistency issue with needlejet combat mechanics
* When scrambled interceptors are defending other units from attack the combat is resolved weapon-to-weapon, but when interceptors initiate fight on their own turn the combat is resolved weapon-to-armor. This biases combat towards overpowered Copters because they can retreat after their attack.

Insufficient checks when AI trades energy for new techs (noted by John-Green223, fixed by Thinker)
* When selecting "urgent need of energy credits" > "some of my valuable research data" dialog path, AI will sometimes trade energy credits that it doesn't have for new techs, thus ending up with a negative balance.

Captured bases switch previous owner faction in an unsual way (fixed by Thinker)
* If faction B captures base from A, and the same base is immediately captured by faction C, it will be treated as previously owned by faction B. This is usually not consistent since it bypasses the turns required (up to 50) for assimilating captured bases.

**Comment by ghost** [comment]
Morale calculation flaws:

- The SMAC manual claims that the Headquarters gives +1 morale on defense, but it does not (p. 104). Implementing this only requires a adding a facility check and could marginally help the AI resist early attacks.
- Drone riots in a unit's home base give -1 morale (undocumented effect AFAIK), but the logic is flawed and ends up messing up other morale effects (see below).
- The classic Children's Creche bug, which is replicated by the Brood Pit. The SMAC manual simply states "All negative morale effects are cancelled for units in base square; instead such units receive a +1 morale modifier." (p. 166) but this is very inaccurate and the logic doesn't work, either. Native units also benefit from both the Creche and Brood Pit in some ways but not others. I wrote up a long treatment if you look at alphacentauri2 (dot) info/wiki/Morale_(Advanced) that describes exactly what happens.

OpenSMACX has various fixes in the code for all these issues, but I'm not sure if they're complete or fully tested. It appears that some of the wackier Creche/Pit effects are retained rather than the flat morale increase. You can also see the original code as it exists in earlier versions.

- [get_basic_defense](https://github.com/b-casey/OpenSMACX/blob/master/src/veh.cpp#L385)
- [morale_veh](https://github.com/b-casey/OpenSMACX/blob/58e36e0bc86a9e050c1cab2d737f7108834ba7c9/src/veh.cpp#L2613)

**Comment by DBlask** [comment]
There's a bug that has bedevilled me from the beginning - a freeze that happens during the player's turn when an automated artillery is near an enemy aircraft.  It doesn't happen with AI artillery, just automated player artillery.  I think it's trying to attack, and of course it's not able to.  I recently had it happen with an automated foil as well.  I haven't played for a few years, but I don't remember that happening in the past

**Reply by induktio** [reply]
Probably that issue has not been reported before. You might have to post some savegame on issues where the bug happens and maybe include more details.
